### PR TITLE
LCCP preparation for CPC

### PR DESCRIPTION
**Pull request by mschoeler**
Changes to LCCP algorithm to prepare it as a base for CPC.
Small corrections to existing code, too.

### CODE REVIEW

File: segmentation/include/pcl/segmentation/lccp_segmentation.h [hunk 1]
```diff
@@ -61,9 +61,18 @@ namespace pcl
...
     {
+      /** \brief Desribes the difference of normals of the two supervoxels being connected*/
+      float normal_difference;
+      
+      /** \brief Desribes if a connection is convex or concave*/
       bool is_convex;
+      
+      /** \brief Desribes if a connection is valid for the segment growing. Usually convex connections are and concave connection are not. Due to k-concavity a convex connection can be invalidated*/
+      bool is_valid;
+      bool used_for_cutting;
```
Comment (taketwo) [inline segmentation/include/pcl/segmentation/lccp_segmentation.h]: Missing description here. Also missing `c` in "Desribes" above.

File: segmentation/include/pcl/segmentation/impl/lccp_segmentation.hpp [hunk 2]
```diff
@@ -397,7 +414,7 @@ pcl::LCCPSegmentation<PointT>::applyKconvexity (unsigned int k_arg)
 
       // Check k convexity
       if (kcount < k_arg)
-        (sv_adjacency_list_)[*edge_itr].is_convex = false;
+        (sv_adjacency_list_)[*edge_itr].is_valid= false;
```
Comment (taketwo) [inline segmentation/include/pcl/segmentation/impl/lccp_segmentation.hpp]: Missing space

File: segmentation/include/pcl/segmentation/impl/lccp_segmentation.hpp [hunk 3]
```diff
@@ -285,6 +285,24 @@ pcl::LCCPSegmentation<PointT>::segment (std::map<uint32_t, typename pcl::Supervo
...
+
+template <typename PointT> void
+pcl::LCCPSegmentation<PointT>::doGrouping ()
+{
+  // clear the processed_ map
+  seg_label_to_sv_list_map_.clear ();
+  for (typename std::map<uint32_t, typename pcl::Supervoxel<PointT>::Ptr>::iterator svlabel_itr = sv_label_to_supervoxel_map_.begin ();
+      svlabel_itr != sv_label_to_supervoxel_map_.end (); ++svlabel_itr)
+  {
+    const uint32_t sv_label = svlabel_itr->first;
```
Comment (taketwo) [inline segmentation/include/pcl/segmentation/impl/lccp_segmentation.hpp]: Maybe `const uint32_t&`? No need to copy.

**Comment by taketwo** [comment]
Thanks! Looks good, please fix the small issues and I'll merge.

**Comment by mschoeler** [comment]
ok done

File: segmentation/include/pcl/segmentation/impl/lccp_segmentation.hpp (hunk 3 above)
Comment (mschoeler) [inline segmentation/include/pcl/segmentation/impl/lccp_segmentation.hpp]: Does it make a difference for basic types?

Comment (taketwo) [inline segmentation/include/pcl/segmentation/impl/lccp_segmentation.hpp]: I think that const reference is treated as an "alias" by the compiler, and therefore incurs no runtime penalty. A const variable is not just an "alias", it actually has memory allocated for it and hence an address. So creating such will incur copying. Unless (as in a trivial case like ours) it is optimized.

But I have to admit that these are just my speculations. I wasn't able to google any proof. The only relevant discussion is [this](http://stackoverflow.com/questions/2641907/do-variable-references-alias-incure-runtime-cost), however the references are not const in that case.

So... do as you prefer, it's a tiny issue anyway.

**Comment by mschoeler** [comment]
ok changed it in several locations with uint32_t.

Cheers

**Comment by mschoeler** [comment]
mmh, do not know if this makes sense in all the locations I changed it. Maybe I can get your opinion on that :)

**Comment by taketwo** [comment]
For me all of them make sense, at least this is exactly what I normally do myself :)

**Pull request merged**